### ISSUE TITLE
nvmf_tgt hits heap-use-after-free in poll_group_update_subsystem() during hotplug testing

### ISSUE THREAD

**Issue by MattDumm** [spdk/spdk#2055]
I'm seeing a heap-use-after-free abort reported by asan, while doing hotplug tests where drives get removed and re-added a bunch of times against an nvmf target. The nvmf target code is based on spdk 20.10 but includes some local changes.

After a bunch of hotplugging, asan sometimes catches a heap-use-after-free dereference of a qpair pointer for an admin_qpair that appears to have been recently destroyed but is still associated with its ctrlr. This is done in poll_group_update_subsystem(), using an rqpair that was already freed by nvmf_rdma_qpair_destroy().
```
            if (ns_changed) {
                    TAILQ_FOREACH(ctrlr, &subsystem->ctrlrs, link) {
                            if (ctrlr->admin_qpair->group == group) {             ###  <<<<< the use-after free
                                    nvmf_ctrlr_async_event_ns_notice(ctrlr);
                            }
                    }
            }
```
The ctrlr has a stale pointer to an admin_qpair which has been destroyed and whose memory was freed. The subsystem resume found that ctrlr in subsystem->ctrlrs, and tried to use that bad admin_qpair pointer.

It looks like the ctrlr should have been destroyed and removed from subsystem->ctrlrs, since its admin_qpair went away. However, in this trace I noted that the admin_qpair didn't seem to be the last qpair deleted from the ctrlr -- another (IO) qpair appears to still be associated with the ctrlr and deleted shortly after. When that final qpair is deleted it calls for destruction of the ctrlr, but before the message to clean up the controller gets processed, there's a subsystem state change, and the ctrlr is still in subsystem->ctrlrs, so it tries to use it.

Some log excerpts below. Line numbers are off from the base 20.10 code due to local mods and added debug logs.

** LOG OUTPUT:
```
# The controller and admin qpair in question:
    cntlid=0x0010  (ptr to ctrlr= 0x622000096100)
    qid=0x0000     (ptr to qpair= 0x6120000c9c40)

* disconnect, and start destroying the admin qpair:
2021-07-22 14:40:12.825 MST {3210484} I {nvmf_tgt_appasan_30164} rdma.c:2825:nvmf_rdma_disconnect !!! disconnect qpair: cntlid=0x0010, qid=0x0000
2021-07-22 14:40:12.825 MST {3210484} I {nvmf_tgt_appasan_30164} nvmf.c:938:_nvmf_qpair_destroy !!! cntlid=0x0010, qid=0x0000
2021-07-22 14:40:12.825 MST {3210484} I {nvmf_tgt_appasan_30164} nvmf.c:941:_nvmf_qpair_destroy XXXX !!! cntlid=0x0010, qid=0x0000 (qpair= 0x6120000c9c40, qpair->ctrlr= 0x622000096100)
2021-07-22 14:40:12.825 MST {3210484} I {nvmf_tgt_appasan_30164} nvmf.c:1030:spdk_nvmf_qpair_disconnect !!! Can't destroy qpair due to outstanding IOs: cntlid=0x0010, qid=0x0000
2021-07-22 14:40:12.825 MST {3210484} E {nvmf_tgt_appasan_30164} nvmf.c:880:_nvmf_ctrlr_free_from_qpair XXXX we were not the last qpair (qpair= 0x6120000c9c40) on this ctrlr...
2021-07-22 14:40:12.825 MST {3210484} E {nvmf_tgt_appasan_30164} nvmf.c:883:_nvmf_ctrlr_free_from_qpair XXXX sending thread msg to _nvmf_transport_qpair_fini() for qpair= 0x6120000c9c40 qpnum= 0


# later, when admin qpair's is drained, we can destroy our rqpair 0x6120000c9c40 (qid 0)... But right at about that same time:
#          - we get a pause req for subsys in question, and
#         - qid 1 disconnect/destroy requested

* state change for our subsys (to state 4 == PAUSED):
2021-07-22 14:40:12.898 MST {3210484} E {nvmf_tgt_appasan_30164} subsystem.c:486:subsystem_state_change_on_pg XXXX stat_chg_on_pg: subsys= 0x614000013440, req_state= 4

* our nvmf_rdma_destroy_drained_qpair() happens, interleaved w/ the pause messages:
2021-07-22 14:40:12.898 MST {3210484} E {nvmf_tgt_appasan_30164} rdma.c:2797:nvmf_rdma_destroy_drained_qpair XXXX 3 destroy rqpair= 0x6120000c9c40, QP#= 0.
2021-07-22 14:40:12.898 MST {3210484} E {nvmf_tgt_appasan_30164} subsystem.c:499:subsystem_state_change_on_pg XXXX PAUSE requested for subsys= 0x614000013440, req_state= 4
2021-07-22 14:40:12.898 MST {3210484} E {nvmf_tgt_appasan_30164} rdma.c:838:nvmf_rdma_qpair_destroy XXXX destroying rpqair= 0x6120000c9c40... rqpair->qpair.state= 3, rqpair->qpair.ctrlr= (nil)
2021-07-22 14:40:12.898 MST {3210484} E {nvmf_tgt_appasan_30164} nvmf.c:1422:nvmf_poll_group_pause_subsystem XXXX pausing subsys= 0x614000013440 ...
  ##  that pause req state machine loops/repeats a few times, during which the other qpair on this controller (qid=0x0001) gets destroyed:
2021-07-22 14:40:12.898 MST {3210484} E {nvmf_tgt_appasan_30164} nvmf.c:1422:nvmf_poll_group_pause_subsystem XXXX pausing subsys= 0x614000013440 ...
2021-07-22 14:40:12.898 MST {3210484} I {nvmf_tgt_appasan_30164} nvmf.c:938:_nvmf_qpair_destroy !!! cntlid=0x0010, qid=0x0001
2021-07-22 14:40:12.898 MST {3210484} E {nvmf_tgt_appasan_30164} subsystem.c:486:subsystem_state_change_on_pg XXXX stat_chg_on_pg: subsys= 0x614000013440, req_state= 4

* qpair_destroy call for the other one, qid=1:
2021-07-22 14:40:12.898 MST {3210484} I {nvmf_tgt_appasan_30164} nvmf.c:941:_nvmf_qpair_destroy XXXX !!! cntlid=0x0010, qid=0x0001 (qpair= 0x61200005c440, qpair->ctrlr= 0x622000096100)
2021-07-22 14:40:12.898 MST {3210484} E {nvmf_tgt_appasan_30164} subsystem.c:499:subsystem_state_change_on_pg XXXX PAUSE requested for subsys= 0x614000013440, req_state= 4
2021-07-22 14:40:12.898 MST {3210484} E {nvmf_tgt_appasan_30164} nvmf.c:1422:nvmf_poll_group_pause_subsystem XXXX pausing subsys= 0x614000013440 ...
2021-07-22 14:40:12.898 MST {3210484} E {nvmf_tgt_appasan_30164} subsystem.c:486:subsystem_state_change_on_pg XXXX stat_chg_on_pg: subsys= 0x614000013440, req_state= 4
2021-07-22 14:40:12.898 MST {3210484} E {nvmf_tgt_appasan_30164} subsystem.c:499:subsystem_state_change_on_pg XXXX PAUSE requested for subsys= 0x614000013440, req_state= 4
2021-07-22 14:40:12.898 MST {3210484} E {nvmf_tgt_appasan_30164} nvmf.c:1422:nvmf_poll_group_pause_subsystem XXXX pausing subsys= 0x614000013440 ...
...
2021-07-22 14:40:12.898 MST {3210484} E {nvmf_tgt_appasan_30164} subsystem.c:450:subsystem_state_change_done XXXX state chg done - subsys= 0x614000013440, req'd_state= 4

* finishing up the qpair destroy of qid 0 here, and freeing it:
2021-07-22 14:40:12.907 MST {3210484} E {nvmf_tgt_appasan_30164} rdma.c:912:nvmf_rdma_qpair_destroy XXXX destroying rpqair= 0x6120000c9c40 but rqpair->qpair.ctrlr is null.

* then later, finishing up qpair destroy of qid 1
2021-07-22 14:40:12.907 MST {3210484} E {nvmf_tgt_appasan_30164} nvmf.c:869:_nvmf_ctrlr_free_from_qpair XXXX sending thread msg to _destruct ctrlr= 0x622000096100 ctrlrid= 0x0010 from subsys 0x614000013440
2021-07-22 14:40:12.907 MST {3210484} E {nvmf_tgt_appasan_30164} nvmf.c:872:_nvmf_ctrlr_free_from_qpair XXXX we were the last qpair (qpair= 0x61200005c440) on this ctrlr, which is being destroyed. Set our qpair->ctrlr to NULL.
2021-07-22 14:40:12.907 MST {3210484} E {nvmf_tgt_appasan_30164} nvmf.c:883:_nvmf_ctrlr_free_from_qpair XXXX sending thread msg to _nvmf_transport_qpair_fini() for qpair= 0x61200005c440 qpnum= 1


* then, before _nvmf_ctrlr_destruct() happens, another subsys state change req for state 2 (ACTIVE) on that same subsystem:
2021-07-22 14:40:12.907 MST {3210484} E {nvmf_tgt_appasan_30164} nvmf.c:1460:nvmf_poll_group_resume_subsystem XXXX resuming subsys= 0x614000013440 ...
2021-07-22 14:40:12.907 MST {3210484} E {nvmf_tgt_appasan_30164} nvmf.c:1247:poll_group_update_subsystem XXXX ctrlr= 0x622000096100, ctrlr->admin_pair= 0x6120000c9c40, about to deref ctrlr->admin_qpair->group!

##   -> the _nvmf_ctrlr_destruct() hadn't happened yet, and the ctrlr was still in its subsystem->ctrlrs list, but didn't have a valid admin_qpair pointer.
##   -> and that deref of ctrlr->admin_qpair(0x6120000c9c40)->group is the deref of freed mem, leading to the asan's abort
```

** ASAN OUTPUT:
```
=================================================================
==3210484==ERROR: AddressSanitizer: heap-use-after-free on address 0x6120000c9c68 at pc 0x00000050e3ce bp 0x7f88c4ef8370 sp 0x7f88c4ef8360
READ of size 8 at 0x6120000c9c68 thread T4 (reactor_1)
...
    #0 0x50e3cd in poll_group_update_subsystem /home/vagrant/.cache/bazel/_bazel_vagrant/11d675efbfe2691e2879cf3ac8b9ddc2/sandbox/linux-sandbox/10/execroot/__main__/bazel-out/k8-fastbuildrelease/bin/external/spdk_archived/spdklib.build_tmpdir/lib/nvmf/nvmf.c:1251
    #1 0x5127b2 in nvmf_poll_group_resume_subsystem /home/vagrant/.cache/bazel/_bazel_vagrant/11d675efbfe2691e2879cf3ac8b9ddc2/sandbox/linux-sandbox/10/execroot/__main__/bazel-out/k8-fastbuildrelease/bin/external/spdk_archived/spdklib.build_tmpdir/lib/nvmf/nvmf.c:1470
    #2 0x5b7391 in msg_queue_run_batch /home/vagrant/.cache/bazel/_bazel_vagrant/11d675efbfe2691e2879cf3ac8b9ddc2/sandbox/linux-sandbox/10/execroot/__main__/bazel-out/k8-fastbuildrelease/bin/external/spdk_archived/spdklib.build_tmpdir/lib/thread/thread.c:505
    #3 0x5b7391 in thread_poll /home/vagrant/.cache/bazel/_bazel_vagrant/11d675efbfe2691e2879cf3ac8b9ddc2/sandbox/linux-sandbox/10/execroot/__main__/bazel-out/k8-fastbuildrelease/bin/external/spdk_archived/spdklib.build_tmpdir/lib/thread/thread.c:581
    #4 0x5b7391 in spdk_thread_poll /home/vagrant/.cache/bazel/_bazel_vagrant/11d675efbfe2691e2879cf3ac8b9ddc2/sandbox/linux-sandbox/10/execroot/__main__/bazel-out/k8-fastbuildrelease/bin/external/spdk_archived/spdklib.build_tmpdir/lib/thread/thread.c:689
    #5 0x603a4b in _reactor_run /home/vagrant/.cache/bazel/_bazel_vagrant/11d675efbfe2691e2879cf3ac8b9ddc2/sandbox/linux-sandbox/10/execroot/__main__/bazel-out/k8-fastbuildrelease/bin/external/spdk_archived/spdklib.build_tmpdir/lib/event/reactor.c:326
    #6 0x603a4b in reactor_run /home/vagrant/.cache/bazel/_bazel_vagrant/11d675efbfe2691e2879cf3ac8b9ddc2/sandbox/linux-sandbox/10/execroot/__main__/bazel-out/k8-fastbuildrelease/bin/external/spdk_archived/spdklib.build_tmpdir/lib/event/reactor.c:382
    #7 0x421979 in eal_thread_loop.cold.0 (/root/mattd/nvmf_tgt_appasan_30164+0x421979)
    #8 0x7f88cc2312dd in start_thread (/lib64/libpthread.so.0+0x82dd)
    #9 0x7f88ca929e82 in __GI___clone (/lib64/libc.so.6+0xfbe82)

0x6120000c9c68 is located 40 bytes inside of 296-byte region [0x6120000c9c40,0x6120000c9d68)
freed by thread T4 (reactor_1) here:
    #0 0x7f88cd0117b0 in __interceptor_free (/lib64/libasan.so.5+0xef7b0)
    #1 0x539a3b in nvmf_rdma_qpair_destroy /home/vagrant/.cache/bazel/_bazel_vagrant/11d675efbfe2691e2879cf3ac8b9ddc2/sandbox/linux-sandbox/10/execroot/__main__/bazel-out/k8-fastbuildrelease/bin/external/spdk_archived/spdklib.build_tmpdir/lib/nvmf/rdma.c:925
    #2 0x52921a in nvmf_rdma_qpair_process_ibv_event /home/vagrant/.cache/bazel/_bazel_vagrant/11d675efbfe2691e2879cf3ac8b9ddc2/sandbox/linux-sandbox/10/execroot/__main__/bazel-out/k8-fastbuildrelease/bin/external/spdk_archived/spdklib.build_tmpdir/lib/nvmf/rdma.c:3041
    #3 0x5b7391 in msg_queue_run_batch /home/vagrant/.cache/bazel/_bazel_vagrant/11d675efbfe2691e2879cf3ac8b9ddc2/sandbox/linux-sandbox/10/execroot/__main__/bazel-out/k8-fastbuildrelease/bin/external/spdk_archived/spdklib.build_tmpdir/lib/thread/thread.c:505
    #4 0x5b7391 in thread_poll /home/vagrant/.cache/bazel/_bazel_vagrant/11d675efbfe2691e2879cf3ac8b9ddc2/sandbox/linux-sandbox/10/execroot/__main__/bazel-out/k8-fastbuildrelease/bin/external/spdk_archived/spdklib.build_tmpdir/lib/thread/thread.c:581
    #5 0x5b7391 in spdk_thread_poll /home/vagrant/.cache/bazel/_bazel_vagrant/11d675efbfe2691e2879cf3ac8b9ddc2/sandbox/linux-sandbox/10/execroot/__main__/bazel-out/k8-fastbuildrelease/bin/external/spdk_archived/spdklib.build_tmpdir/lib/thread/thread.c:689
    #6 0x603a4b in _reactor_run /home/vagrant/.cache/bazel/_bazel_vagrant/11d675efbfe2691e2879cf3ac8b9ddc2/sandbox/linux-sandbox/10/execroot/__main__/bazel-out/k8-fastbuildrelease/bin/external/spdk_archived/spdklib.build_tmpdir/lib/event/reactor.c:326
    #7 0x603a4b in reactor_run /home/vagrant/.cache/bazel/_bazel_vagrant/11d675efbfe2691e2879cf3ac8b9ddc2/sandbox/linux-sandbox/10/execroot/__main__/bazel-out/k8-fastbuildrelease/bin/external/spdk_archived/spdklib.build_tmpdir/lib/event/reactor.c:382
    #8 0x421979 in eal_thread_loop.cold.0 (/root/mattd/nvmf_tgt_appasan_30164+0x421979)
```

## Expected Behavior

Should be able to remove and re-add drives without seeing use-after free. Ctrlr should not be used by poll_group_update_subsystem() after admin qpair is gone.

## Current Behavior

poll_group_update_subsystem() accesses bad admin qpair pointer, aborts with heap-use-after-free.

## Steps to Reproduce

Encountered this while performing random hotplug sequences in which we stressfully remove and re-add drives.

## Context (Environment including OS version, SPDK version, etc.)

SPDK 20.10
CentOS Linux release 8.0.1905 (Core)
Linux localhost.localdomain 4.18.0 #1 SMP Tue May 11 11:32:13 PDT 2021 x86_64 x86_64 x86_64 GNU/Linux

**Comment by AlekseyMarchuk** [comment]
Hi @MattDumm this particular issue should already be fixed by Jim Harris in this patchset: https://review.spdk.io/gerrit/c/spdk/spdk/+/8570. It fixes a bit different scenario (admin qpair is not assigned when poll group is updated) but I believe it should cover controller disconnect case as well. The fix is included into SPDK 21.07 which is released this week. Could you try to run your test with this fix?

**Comment by mhae** [comment]
@AlekseyMarchuk , thanks for the pointer. However, this fix won't apply here because the admin_queue was set before and now contains an already freed pointer in our scenario.
When we free the RDMA admin qpair, we should set the ctrlr->admin_qpair to NULL. We've tried that but run into other issues where the ctrlr may not be set yet (Matt has more details).

**Comment by mhae** [comment]
@AlekseyMarchuk , I this is tricky to reproduce. We do actual drive hot plugs while an (NVMF) initiator is connected. It doesn't happen all the time and needs to run for quite some time. We have a whole infrastructure to test this with actual HW.

The bottom line is that an initiator can disconnect (see nvmf_rdma_qpair_process_ibv_event in above output) while the subsystem is paused due to the drive pull. As part of the subsystem pause, we try to send an AEN back to the initiator but the admin qpair is no longer valid.
I think we need to find a way to set the ctrlr->admin_qpair to NULL when we free it.

We are happy to try out anything you suggest. It's easy for us to run these tests.

**Comment by MattDumm** [comment]
@AlekseyMarchuk , thanks for looking at this one. As @mhae said, the null check for the ctrlr->admin_qpair pointer (https://review.spdk.io/gerrit/c/spdk/spdk/+/8570) does not cover our current problem. We would also need ctlrl->admin_qpair to be set to null when the admin qpair is destroyed so we don't accidentally try to use it again in a situation like a subsystem update.

I tinkered with changing nvmf_rdma_qpair_destroy() for admin qpairs to have them set the parent ctrlr's admin qpair to null, adding this before the "free(rqpair);" at the function's end:
```
    if (rqpair->qpair.ctrlr->admin_qpair == &rqpair->qpair) {
        rqpair->qpair.ctrlr->admin_qpair = NULL;
    }
```
I got asan abort trouble with that approach - first with the discovery controller and then in some other situations where either the qpair or ctrlr state wasn't appropriate to be doing that there. I wasn't sure under what conditions you can trust qpair.ctrlr to be a safe, init'd, non-freed pointer so I didn't go any further with that approach. Maybe that difficulty could be worked out in nvmf_rdma_qpair_destroy() and we could null that pointer there, then null-check it in poll_group_update_subsystem() using the fix you referred to.

But maybe a completely different approach is needed, like somehow closing the race window between destruction of the admin_qpair and destruction of the ctrlr and its removal from the subsystem->ctrlrs list. It looks like controller destruct is called for only after the final qpair is destroyed, and that final qpair won't always be the admin qpair. And even if the admin qpair is the last one destroyed, the path from there to controller destroy is asynchronous and it seems like a subsystem update could sneak in that window unless some other guard mechanism is used.

**Comment by AlekseyMarchuk** [comment]
Hi @MattDumm @mhae I decided to start from simple solution which you implemented first - set admin_qpair pointer to NULL. I think it should work fine since we set a pointer to NULL and call poll_group_update_subsystem on the same thread. Could you try this patch https://review.spdk.io/gerrit/c/spdk/spdk/+/8963 together with https://review.spdk.io/gerrit/c/spdk/spdk/+/8570 ?

**Comment by MattDumm** [comment]
Hi @AlekseyMarchuk - Sorry for the delayed response. I was away from work for a couple days.
I'll take a look at this and try the patch as soon as I can. My test machine is down for maintenance today, so it will be later this week before I can get back to you.

**Comment by MattDumm** [comment]
@AlekseyMarchuk - I haven't forgotten this, but still having issues w/ my hardware. I will try your suggested fixes next week and let you know the results.

**Comment by AlekseyMarchuk** [comment]
Thanks @MattDumm no worries)

**Comment by MattDumm** [comment]
@AlekseyMarchuk - Sorry for the delays. I finally got my equipment back working and was able to do some testing with and without the two patches you referred to above (8963 and 8570). In the days since I last worked with this, we migrated our source code from a 20.10-base to a 21.04 base. So this followup testing is on the (modified) 21.04 spdk code base.

Results so far, after couple hours of beating on the patched and unpatched versions of the nvmf_tgt app with hotplug remove/add of several devices at a time (using address sanitizer):

- There was a repro of the use-after-free abort one time on the *unfixed* version of the nvmf_tgt.
- I did not hit that problem on the fixed version of the nvmf_tgt.

This supports the idea that the fixes probably address this problem, but more hours/tests are needed if we want to feel statistically confident about that statement.

However, in the fixed version there was one encounter with a different use-after-free abort, this time in _nvmf_transport_qpair_fini_complete(), in the new lines of code from https://review.spdk.io/gerrit/c/spdk/spdk/+/8963.
In the line "ctrlr->admin_qpair = NULL;"  -- the ctrlr pointer in this case was the free'd memory being deref'd.

In _nvmf_qpair_destroy() I notice that the code checks for a null ctrlr before calling this:
     "nvmf_transport_qpair_fini(qpair, _nvmf_transport_qpair_fini_complete, qpair_ctx);"

This seems to imply that a NULL ctrlr is something that can happen and it's not always safe for nvmf_transport_qpair_fini_complete() to use the ctrlr pointer. Maybe a null check would avoid this situation, not sure. More investigation needed.

**Comment by MattDumm** [comment]
This fix is subject to a race condition - the ctrlr pointer that we want to use in _nvmf_transport_qpair_fini_complete() may have already been freed.
In _nvmf_ctrlr_free_from_qpair(), if the qpair being removed is the last one in ctrlr->qpair_mask, it first sends a _nvmf_ctrlr_destruct() message (which will free the ctrlr ptr), then sends a _nvmf_transport_qpair_fini() message, which will result in the _nvmf_transport_qpair_fini_complete() callback. If the destruct completes first then the callback is going to have a bad time.

If we want to use the ctrlr pointer from that callback, would it be wise reorder how _nvmf_ctrlr_free_from_qpair() does things, to make sure that the qpair_fini was complete before the destruct message was sent? I'm not sure if that's the wrong direction to go though -- we never needed the ctrlr pointer in that callback before so I'm not sure if it's appropriate to reorder things to make it available there. (or if that would cause some other problem).

**Comment by AlekseyMarchuk** [comment]
Thanks for your feedback @MattDumm I was on vacation last week, going to continue working on this issue

**Comment by AlekseyMarchuk** [comment]
@MattDumm I updated this patch https://review.spdk.io/gerrit/c/spdk/spdk/+/8963
In the latest version I a bit reworked how controller is destroyed - previously we started to destroy the controller when the last qpair was active (just before destroy the qpair) but now we destroy the controller when the last qpair finishes the disconnect process. It should remove the raise condition that you mentioned

**Comment by MattDumm** [comment]
Hi @AlekseyMarchuk -
I'm having some trouble using the fix. I get a crash toward the end of init, in spdk_nvmf_poll_group_remove() where it is trying to use the qpair pointer for the "TAILQ_FOREACH(tgroup, &qpair->group->tgroups, link)". The qpair pointer at that point is bad, and it segfaults.

#0  0x000000000051b752 in spdk_nvmf_poll_group_remove (qpair=qpair@entry=0x6120000616c0) at nvmf.c:1038
#1  0x000000000051bcf2 in _nvmf_qpair_destroy (ctx=0x604000012490, status=<optimized out>) at nvmf.c:1088
#2  0x000000000051beda in spdk_nvmf_qpair_disconnect (qpair=0x6120000616c0, cb_fn=0x0, ctx=0x0) at nvmf.c:1176
#3  0x000000000051c624 in _nvmf_qpair_disconnect_msg (ctx=0x60400002a190) at nvmf.c:1097
#4  0x00000000005d744a in msg_queue_run_batch (max_msgs=<optimized out>, thread=0x619000011880) at thread.c:544
#5  thread_poll (thread=thread@entry=0x619000011880,
    max_msgs=<error reading variable: Cannot access memory at address 0x7f51eb6f65f0>,
    now=now@entry=4225468885923612) at thread.c:623
#6  0x00000000005d91c6 in spdk_thread_poll (thread=thread@entry=0x619000011880, max_msgs=max_msgs@entry=0,
    now=4225468885923612) at thread.c:733
#7  0x000000000062f1f0 in _reactor_run (reactor=0x620000000600) at reactor.c:906
#8  reactor_run (arg=0x620000000600) at reactor.c:944
#9  0x00000000004218ae in eal_thread_loop.cold ()
    at bazel-out/k8-fastbuildrelease/bin/external/easyloggingpp/_virtual_includes/easyloggingpp/easylogging/easylogging++.h:3293
#10 0x00007f51f326f2de in start_thread () from /lib64/libpthread.so.0
#11 0x00007f51f1967e83 in clone () from /lib64/libc.so.6

It's possible that there are other changes between current master and our 21.04-based branch which could affect this. Also possible that I merged your fix wrongly. I will double check these and try to get better details / debugging and let you know what I find.

**Comment by MattDumm** [comment]
A quick update: I found the issue causing my crash, and it was a merge problem which happened because our 21.04 base has diverged from upstream master. Two noteworthy commits were b503ef4fa0 and 103f5c62f5, both impacting the same area we're looking at for this fix...  it left my 21.04 "fixed" version with 2 calls to spdk_nvmf_poll_group_remove() in our path instead of one, and crashed there.
When I straightened out the crash problem I ran a bunch of hotplug through the new code and didn't see that issue yet. I'll do more testing hours and confirm this. (unfortunately, repro of the original problem is rare and can take a lot of hours/test iterations).
Thanks!

**Comment by AlekseyMarchuk** [comment]
@MattDumm verified [this ](https://review.spdk.io/gerrit/c/spdk/spdk/+/8963) patch and set +1. It is ready for review

**Comment by MattDumm** [comment]
Yes, code looks good to me. The issue no longer happened with the change incorporated.

**Comment by MattDumm** [comment]
Any ETA for when this will get final reviews and be merged into the code? It all looks good from my side.

**Comment by MattDumm** [comment]
Running recently without using asan, I'm now seeing that there are some shutdown issues with the code.

I don't have the root cause figured out but in my code version (in which this fix was backported to a 21.04 base with a few modifications, and thus may not behave exactly like your code based on current master) I am seeing two different crash stacks sometimes during shutdown

 CRASHED!
 === Backtrace ===
    @     0x7f40bbc26790  (unknown)
    @           0x574130  spdk_ring_enqueue
    @           0x54fedf  spdk_thread_send_msg
    @           0x50b923  _nvmf_ctrlr_free_from_qpair
    @           0x54d928  thread_poll
    @           0x54e5c0  spdk_thread_poll
    @           0x56e8a1  reactor_run
    @           0x415f03  eal_thread_loop.cold.0
    @     0x7f40bd5f22de  start_thread
    @     0x7f40bbceae83  __GI___clone
    @              (nil)  (unknown)


 CRASHED!
 === Backtrace ===
    @     0x7f95dec22790  (unknown)
    @           0x50b0c8  nvmf_subsystem_remove_ctrlr
    @           0x501f1b  nvmf_ctrlr_destruct
    @           0x54d928  thread_poll
    @           0x54e5c0  spdk_thread_poll
    @           0x56e8a1  reactor_run
    @           0x56ec67  spdk_reactors_start
    @           0x56baec  spdk_app_start
    @           0x424afb  _ZZN6nvgrid15NvmfTargetStartEiPPcENKUlvE2_clEv
    @           0x426ddc  _ZSt13__invoke_implIvZN6nvgrid15NvmfTargetStartEiPPcEUlvE2_JEET_St14__invoke_otherOT0_DpOT1_
    @           0x426618  _ZSt8__invokeIZN6nvgrid15NvmfTargetStartEiPPcEUlvE2_JEENSt15__invoke_resultIT_JDpT0_EE4typeEOS5_DpOS6_
    @           0x427a26  _ZNSt6thread8_InvokerISt5tupleIJZN6nvgrid15NvmfTargetStartEiPPcEUlvE2_EEE9_M_invokeIJLm0EEEEDTcl8__invokespcl10_S_declvalIXT_EEEEESt12_Index_tupleIJXspT_EEE
    @           0x4279e2  _ZNSt6thread8_InvokerISt5tupleIJZN6nvgrid15NvmfTargetStartEiPPcEUlvE2_EEEclEv
    @           0x4279a6  _ZNSt6thread11_State_implINS_8_InvokerISt5tupleIJZN6nvgrid15NvmfTargetStartEiPPcEUlvE2_EEEEE6_M_runEv
    @     0x7f95df609a33  (unknown)
    @     0x7f95e05ee2de  start_thread
    @     0x7f95dece6e83  __GI___clone
    @              (nil)  (unknown)

I will work to get more details. But for now I just wanted to let you know I've hit some shutdown problems.

**Comment by AlekseyMarchuk** [comment]
Hi @MattDumm is it possible to get coredump with variables?
I have a question about your app - are you using a standard nvmf target or your own application? It looks like spdk threads are already destroyed when some controllers are still exist.

**Comment by MattDumm** [comment]
@AlekseyMarchuk -
About the app: It is based on the nvmf_tgt code from spdk (version 21.04), But it includes some product specific extensions to integrate with our platform, and it has some extra debug messages and a few fixes for a issues that came up in our testing.

Here's a stack trace with more detail, for the 1st stack listed above. (file/line numbers won't line up exactly with your version of nvmf_tgt code but it's basically the same flow, aside from where there have been recent changes since 21.04):

(gdb) bt
#0  0x00007f4c7d96e70f in raise () from /lib64/libc.so.6
#1  0x00007f4c7d958b25 in abort () from /lib64/libc.so.6
#2  0x00000000004a01bd in el::base::utils::abort (status=11, reason="Application has crashed due to [SIGSEGV] signal - [Called el::Helpers::crashAbort(11)]")
    at external/easyloggingpp/easylogging++.cc:125
#3  0x00000000004b12cd in el::Helpers::crashAbort (sig=11, sourceFile=0x5ec5a3 "", line=0) at external/easyloggingpp/easylogging++.cc:3121
#4  0x0000000000499171 in (anonymous namespace)::NvCrashHandler (sig=11) at util/log.cc:140
#5  <signal handler called>
#6  0x0000000000574290 in rte_ring_enqueue_bulk_elem (free_space=0x0, n=1, esize=8, obj_table=0x7f4c7abdf740, r=0x0)
    at /home/vagrant/.cache/bazel/_bazel_vagrant/11d675efbfe2691e2879cf3ac8b9ddc2/sandbox/linux-sandbox/16/execroot/__main__/bazel-out/k8-dbg/bin/external/spdk_archived/spdklib.build_tmpdir/dpdk/build/include/rte_ring_elem.h:572
#7  rte_ring_enqueue_bulk (free_space=0x0, n=1, obj_table=0x7f4c7abdf740, r=0x0)
    at /home/vagrant/.cache/bazel/_bazel_vagrant/11d675efbfe2691e2879cf3ac8b9ddc2/sandbox/linux-sandbox/16/execroot/__main__/bazel-out/k8-dbg/bin/external/spdk_archived/spdklib.build_tmpdir/dpdk/build/include/rte_ring.h:266
#8  spdk_ring_enqueue (ring=0x0, objs=objs@entry=0x7f4c7abdf740, count=count@entry=1, free_space=free_space@entry=0x0) at env.c:428
#9  0x000000000055003f in spdk_thread_send_msg (thread=0x7f4c6c4ca880, fn=fn@entry=0x50baa0 <_nvmf_ctrlr_destruct>, ctx=ctx@entry=0x7f4c64142e50) at thread.c:963
#10 0x000000000050b9e0 in _nvmf_ctrlr_free_from_qpair (ctx=0x7f4c601c43e0) at nvmf.c:950
#11 0x000000000054d9e8 in msg_queue_run_batch (max_msgs=<optimized out>, thread=0x7f4c6c418700) at thread.c:546
#12 thread_poll (thread=thread@entry=0x7f4c6c418700, max_msgs=max_msgs@entry=0, now=now@entry=3412731284671760) at thread.c:625
#13 0x000000000054e680 in spdk_thread_poll (thread=thread@entry=0x7f4c6c418700, max_msgs=max_msgs@entry=0, now=3412731284671760) at thread.c:735
#14 0x000000000056ea01 in _reactor_run (reactor=0x7f4c6c40a300) at reactor.c:906
#15 reactor_run (arg=0x7f4c6c40a300) at reactor.c:944
#16 0x0000000000415f03 in eal_thread_loop.cold ()
    at /home/vagrant/.cache/bazel/_bazel_vagrant/11d675efbfe2691e2879cf3ac8b9ddc2/sandbox/linux-sandbox/16/execroot/__main__/bazel-out/k8-dbg/bin/external/spdk_archived/spdklib.build_tmpdir/dpdk/build/include/rte_mempool.h:676
#17 0x00007f4c7f33a2de in start_thread () from /lib64/libpthread.so.0
#18 0x00007f4c7da32e83 in clone () from /lib64/libc.so.6

In frame 10, _nvmf_ctrlr_free_from_qpair() is doing this:
      spdk_thread_send_msg(ctrlr->subsys->thread, _nvmf_ctrlr_destruct, ctrlr);     # nvmv.c line 950

At that point, ctrlr->subsys looks reasonable but ctrlr->subsys->thread (which is the thread used for the send_msg) looks ugly, like maybe it's freed space:

(gdb) p ctrlr->subsys
$27 = (struct spdk_nvmf_subsystem *) 0x7f4c6c4caa40
(gdb) p *ctrlr->subsys
$28 = {thread = 0x7f4c6c4ca880, id = 1811941584, state = 32588, subtype = SPDK_NVMF_SUBTYPE_NVME, next_cntlid = 1, flags = {allow_any_host = 1 '\001', allow_any_listener = 0 '\000',
    ana_reporting = 0 '\000', reserved = 0 '\000'}, changing_state = false, tgt = 0x7f4c6c4160b0, ns = 0x7f4c6c424450, max_nsid = 1, ctrlrs = {tqh_first = 0x0,
    tqh_last = 0x7f4c6c4caa70}, mutex = {__data = {__lock = 0, __count = 0, __owner = 0, __nusers = 0, __kind = -1, __spins = 0, __elision = 0, __list = {__prev = 0x0,
        __next = 0x0}}, __size = '\000' <repeats 16 times>, "\377\377\377\377", '\000' <repeats 19 times>, __align = 0}, hosts = {tqh_first = 0x0, tqh_last = 0x7f4c6c4caaa8},
  listeners = {tqh_first = 0x0, tqh_last = 0x7f4c6c4caab8}, entries = {tqe_next = 0x0, tqe_prev = 0x0}, sn = '0' <repeats 20 times>,
  mn = "HPE Storage - nvgrid controller\000\000\000\000\000\000\000\000\000", subnqn = "nqn.2019-20.com.hpe:51402ec01492a822.1.bay02", '\000' <repeats 179 times>}
(gdb) p ctrlr->subsys->thread
$29 = (struct spdk_thread *) 0x7f4c6c4ca880
(gdb) p *ctrlr->subsys->thread
$30 = {tsc_last = 139966210488560, stats = {busy_tsc = 139966206183632, idle_tsc = 281479271677954}, active_pollers = {tqh_first = 0x7f4c6c4160b0, tqh_last = 0x7f4c6c41a230},
  timed_pollers = {tqh_first = 0x1, tqh_last = 0x0}, paused_pollers = {tqh_first = 0x7f4c6c4ca8b0, tqh_last = 0x0}, messages = 0x0, msg_fd = -1, msg_cache = {slh_first = 0x0},
  msg_cache_count = 0, critical_msg = 0x0, id = 139966211205352, state = SPDK_THREAD_STATE_RUNNING, pending_unregister_count = 0, io_channels = {tqh_first = 0x7f4c6c4ca8f8,
    tqh_last = 0x0}, tailq = {tqe_next = 0x0, tqe_prev = 0x3030303030303030},
  name = '0' <repeats 12 times>, "\000HPE Storage - nvgrid controller\000\000\000\000\000\000\000\000\000\000nqn.2019-20.com.hpe:51402ec01492a822.1.bay01", '\000' <repeats 158 times>, cpumask = {
    str = '\000' <repeats 23 times>, "\305\001\000\000\000\000\000\000\200\250LlL\177\000\000\320\b\000lL\177\000\000\002\000\000\000\001\000\001\000\260`AlL\177\000\000PDBlL\177\000\000\001", '\000' <repeats 15 times>, "p\252LlL\177", '\000' <repeats 18 times>, "\377\377\377\377", '\000' <repeats 28 times>, "\250\252LlL\177\000\000\000\000\000\000\000\000\000\000\270\252LlL\177", '\000' <repeats 18 times>, '\060' <repeats 20 times>..., cpus = "com.hpe:51402ec01492a822.1.bay02", '\000' <repeats 95 times>}, exit_timeout_tsc = 0,
  in_interrupt = false, fgrp = 0x0, ctx = 0x7f4c6c4cabc0 ""}

Since thread->messages is 0x0, the spdk_ring_enqueue() call blows up (ring = 0x0).

It looks like some pattern leads my code to use of a freed ctrlr->subsys->thread, possibly a problem in the subsys's lifecycle. If you have any thoughts on how that could happen that'd be helpful. I'm going to keep looking at it. And of course, since my code is different from your code, it may only happen in my modified 21.04 code.

**Comment by AlekseyMarchuk** [comment]
Yeah, it looks like thread is already destroyed. I see in nvmf_tgt that we first stop subsystems, than destroy poll groups (and threads) and only than destroy nvmf target (spdk_nvmf_tgt_destroy function which destroys subsystems and controllers if any).
As part of NVMF_TGT_FINI_DESTROY_POLL_GROUPS state we mark a thread as exiting. The thread remains in exiting state until there are registered pollers or io_channels. Once there are no more pollers/io_channels, the thread will be destroyed. There is a chance that some subsystem/controller still exists while the thread is destroyed.
I think that we need to reverse the order - first stop subsystems, than destroy the target and only than destroy poll groups. I'll prepare a patch to change it, will update you soon

**Comment by AlekseyMarchuk** [comment]
Hi @MattDumm I prepared a [patch ](https://review.spdk.io/gerrit/c/spdk/spdk/+/9668 ) which I believe should fix the latest problem. You'll also need to port this [patch](https://github.com/spdk/spdk/commit/97385af196bf271645c7158e97c53cbaf74d2f80#diff-2824b9d6615af855a06a418c93304c3fdf67fb6ccca7c8c3fe89a3aec2c73d28) which makes subsystem destruction asynchronous.

**Comment by MattDumm** [comment]
Thanks @AlekseyMarchuk ! I'll try it out and let you know.

**Comment by AlekseyMarchuk** [comment]
@MattDumm I see that CI fails with the latest patch. Please wait until I fix it

**Comment by MattDumm** [comment]
@AlekseyMarchuk - Understood. I'll be away Wednesday but will check back in Thursday. Thanks again.

**Comment by AlekseyMarchuk** [comment]
@MattDumm the patch passes CI and ready for verification

**Comment by MattDumm** [comment]
@AlekseyMarchuk - Still working on testing on my end. The underlying changes that make subsystem destruction asynchronous are complicating matters but I'll let you know what I find.

**Comment by AlekseyMarchuk** [comment]
This patch 3b1f13e caused CI failures with uring, so it was reverted. I'll continue working on the new version next week

**Comment by AlekseyMarchuk** [comment]
Long time passed, I tried to reproduce failure from CI locally, but this test never failed. When I rebased my patch on top of master, CI passed. I don't know what was the reason of this failure and I would like to highlight that it was specific for uring only. The last version of patch is on review https://review.spdk.io/gerrit/c/spdk/spdk/+/9995
